### DISCUSSION TITLE
Better workflow for local development -> supabase environment

### DISCUSSION THREAD

**Discussion by ericwooley** [supabase/supabase#6846] [Feature Requests]
# Updates 

## May 1st 2023 -  Self hosted Logs

Regarding point 3 below, Supabase Logs is now available for [self-hosting](https://supabase.com/docs/reference/self-hosting-analytics/introduction) as well as [local development](https://supabase.com/docs/guides/getting-started/local-development#enabling-local-logging)

---

💯 for this amazing setup out of the box, with a hosted environment. I really like the overall direction, philosophy, and ease of use of supabase. I intend to make this my new goto for side projects, and experiments. However, as someone starting out pretty fresh with supabase, and having worked on dozens of teams and projects at various startups, I have a few suggestions to help people getting started. These are all suggestions coming from someone who prefers to have a clear vision of how things will scale as the size of the team grows. I make a lot of small projects and apps to see how things go, but I like to set them up as if 100 people will be working on this, so that I can confidently recommend this setup to a client, or colleague. 

1. Having a clear `Local -> Dev -> Staging -> Prod` workflow:
    In my experience, it is very common for teams to work together on a project where everyone can run the project on their local machine, commit the changes, create a pull request to a `dev` branch, have code reviewed, then once merged, have the code deployed to a "dev" environment. Whenever the team is comfortable with dev, the dev branch is merged into `staging`, which has several tests run against it, and if staging passes, then there is a deployment to prod.
    
    Right now, it's very unclear to me as to how to setup this workflow. My current plan is to create a script, which  runs `docker run supabase/pgadmin -schema -diff 'postgres://user:pass@local:5432/diff_source' 'postgres://user:pass@dev:5432/diff_target'`.  to see if there is anything changed. If so, create a new migration file, and push the changes to that file, so that the developer can interact with the supabase studio, and have migrations created, in a reviewable format.

    Ideally, this should happen automatically by triggers in the CLI, similarly to hasura, which generates migrations automatically based on what you do in the UI, because supabase studio has better context then a diff script does, it could generate better names etc... 
    
    The last stage to this setup would be clearer guidance on how to run these migrations. I see lots of references to various diff tools etc... in [this support page](https://supabase.com/blog/2021/03/31/supabase-cli), but no real help on how to actually use the migrations generated. I think the best I can do in this regard, is hack together my own migration tool. Ideally I could just run `supabase migration run <targetdb>`, and the necessary migrations would run, and be recorded in some meta_migrations table, or whatever. It looks like `new` is the only sub command for `supabase migration`, so maybe this is already planned? If hacking together my own migration tool is not necessary, I would love some more documentation on how that is supposed to work. I haven't been able to find anything, yet still see a migrations folder, that makes me think I am missing something. 
    
2. Labeling things as experimental, or a work in progress.
    I have mostly been working on migrations so far, so I'll keep going with that example. 
    
    Given that the only subcommand for migrations is `new`, it seems like maybe this just isn't done being worked on yet? It would be great if the help in the CLI mentioned the WIP nature of the migrations command. The DX is kind of rough, when I see this feature that looks like it's what I want, but I can't really seem to find any good docs on how to use it, only to eventually find that it seems to be more of a WIP. 
    
    As a bonus, to this, if there were a link included in WIPS in the help of the CLI, which directs the user to the roadmap for this feature, so that a developer could have a good sense of the official direction of the CLI. That way the user can hack together whatever they need to, with confidence that thier hack scripts will not conflict with the supabase plan moving forward. 
    
3. Start supabase in the foreground. 

    Right now, running `supabase start`, starts the services, then generates some information about your supabase setup, and then exits. It would be amazing to have a --foreground option, which shows you the logs of the services running, then closes the supabase services when the process exits.
    
4. The local supabase studio seems to be pretty far behind the hosted one in terms of broken links, the functions, log explorer etc.. 

I want to emphasize, I am not trying to criticize the project, I think you guys have done an awesome job, and it's better than anything else out there. I just want to point out some pain points from the perspective of a new user, assessing this for long term projects, and medium to large size teams.

**Comment by adimyth** [comment]
**_Point 1_**:  Hasura cli does a good job at managing postgres schema migrations. In my setup, I have hasura on top of supabase postgres & I use hasura cli to manage both schema migrations & metadata management

**_Point 4_**: Totally agree with ur last point, the local dockerised studio is nowhere close to the hosted ones. The other day I wanted to enable external authentication like google & github, but there was no option for it in the dockerised studio. Finally, added it to the `docker-compose` file & restarted to make it work

**Reply by ericwooley** [reply]
> Point 1: Hasura cli does a good job at managing postgres schema migrations. In my setup, I have hasura on top of supabase postgres & I use hasura cli to manage both schema migrations & metadata management

This is a little off topic, but does that work well? Hasura also has row/column level security, but I'm not sure if thats handled through postgres, or through logic in their gql resolver. Are they compatible with supabase?

**Comment by Ziinc** [comment]
As an update to #4, Supabase Logs is now available for [self-hosting](https://supabase.com/docs/reference/self-hosting-analytics/introduction) as well as [local development](https://supabase.com/docs/guides/getting-started/local-development#enabling-local-logging)
